### PR TITLE
ColorPalette: make sure "key" is unique when iterating over color entries with same value

### PR DESCRIPTION
**Pull request by ciampo**
## What?

While testing a feature in the editor, I noticed that there was an error in the console while opening my personal site's color palette — I realised that the issue was related to the fact that my specific color palette for that specific site has a duplicated color (ie. two separate entries with the same `value` and `name`).

When that happen, it causes 2 issues:

- the aforementioned console error, due to the fact that there are duplicate `key` attributes
- both instances of the color appear as "checked", since they technically are both representing the selected `value`

## Why?

This is an edge case that was never considered before.

## How?

- The `key` issue has been solved by using the array `index` to compute the `key`
- To avoid two instances of the same color appearing as selected, I've arbitrarily decided that only the first matching color will appear as selected. This is not ideal, but IMO it's better than the current behavior.

An alternative idea that came to mind is to filter the list of `colors` and remove duplicates, although that also poses a few challenges:

- it modifies the palette, which could result in an unexpected behavior to the user
- should we consider as duplicate two different entries with the same `value` but a difference `name` ? (note that currently we're not keeping track of the name associated to the `color` that was clicked by the user)

## Testing Instructions

I've added a new Storybook example to showcase the situation described above — you can test the example with and without the changes introduced in this PR to confirm that the fix is working as expected.

### CODE REVIEW

**Comment by marceloaof** [comment]
Probably is a different issue, but it seems similar, so I would like to offer as feedback for your nice work: when creating a white color on the palette, the color value needs to be unnecessarily different from #FFF in order to not resemble the default white color in the palette. The ideal scenario is: the possibility for more than one #FFF color to exist under different names, for different purposes. And being able to select which one to use on each case.
Currently, Twenty-Twenty-Two sets a white color for the background. Two colors appear as checked, and it is not possible to choose which one should really be checked.
image

**Comment by ciampo** [comment]
Hey @marceloaof, what you're flagging is definitely related to the same issue that this PR is trying to mitigate, and it actually highlighted another edge case that I didn't consider (duplicated colors across separate palettes)

- in your example, there are two separate entries in the color palette sharing the same color value (ie. `#ffffff`)
- when clicking one of these two values, the selected color become `#ffffff`
- prior to this PR, the `ColorPalette` component would simply compare each of its entries to the selected value, and if the two vales are the same, it would mark that specific entry as selected. As you can imagine, this means that _all_ color entries with that value will be flagged as selected.

The solution currently proposed in this PR currently doesn't fix this scenario, which makes think that we should be probably looking at a more systematic solution.

---

Here's some reflections:

 - What qualifies as a "duplicate entry"?
   - Entries with the same color `value` but a different `name` should be considered duplicates?
   - Entries with the same color `value` and same color `name` should be considered duplicates?
   - Should we somewhat add an `id` to each entry, guaranteed to be unique to that color entry?
 - If we agree that "duplicate entries" are a problem that needs to be dealt with:
   - We should probably put additional validation check in every place where it's possible to edit such values (I assume both in Gutenberg's UIs, and in the theme.json validation?) so that it's not possible to add a duplicate color entry?
   - We should also agree on the behavior of the `ColorPalette` component in case it receives "duplicate" colors:
     - should the component keep all instances of the duplicate entries, and show them all as selected if they match the currently selected `value` ? (current behavior)
     - should the component filter out and hide all duplicate entries, apart from the first instance?
     - should we look into changing what the component expects as the `value` ? E.g. instead of passing only the color's `value` (eg the rgb or hex code), we could pass the whole color object (which includes `name`, `value`) or even a unique `id` (in case we force each color "value" to have one)
     - any other suggestion?
- How would this translate for gradients?

Personally, I believe that the "safest" way to handle this issue would be to add a unique `id` to each color passed to `ColorPalette`, although I'm not sure about how feasible this is (should these ids be generated automatically? And at what point? Would they be automatically persisted?)

Cc'ing also few folks who may help with this (@mtias @jorgefilipecosta @ntsekouras @andrewserong @aaronrobertshaw @jasmussen )

**Comment by jasmussen** [comment]
Great ticket. I actually ran into this issue a while back, but ended up not ticketing since my specific use case is a bit of a lazy hack. This is a much older GIF, but it shows the issue:
older color bug

Here's what's going on:

* I have two Black colors. "Black" (#000) and "Interface mode black" (also #000)
* The first "Black" is always pure black. But the second one is black only in light mode, but gets inverted in dark mode. 

Like so:

```
html body {
	@media (prefers-color-scheme: dark) {
		--wp--preset--color--interface-mode-black: #fff;
		--wp--preset--color--interface-mode-white: #000;
		--wp--preset--color--interface-mode-gray: #333;
	}
}
```

Yes, it's a bit of a hack to transform the colors like that. But the net result is that I can selectively apply colors on [my site](https://moc.co/) that I know will invert in dark mode, as I traverse the site editor. 

While a hack, it does illustrate one _opportunity_ for theme developers to be creative. For that reason, I would personally compare active colors only on their slug, and not their value. What do you think?

**Comment by ciampo** [comment]
> Yes, it's a bit of a hack to transform the colors like that. But the net result is that I can selectively apply colors on [my site](https://moc.co/) that I know will invert in dark mode, as I traverse the site editor.
> 
> While a hack, it does illustrate one _opportunity_ for theme developers to be creative.

Thank you for sharing additional perspective — your point of view makes a lot of sense!

> For that reason, I would personally compare active colors only on their slug, and not their value. What do you think?

If by `slug` you mean the color's `name` (what in your example were "Black" and "Interface mode black"), using the color's name will only mitigate the problem, but not solve it completely / systematically. This is because I believe it's possible define two color entries that have the same `name` _and_ `color`.

That's why I was suggesting that we consider adding a new unique `id` field to these color objects that is supposed to be unique across the whole list of predefined colors for the site — although I'm aware that this will pose some technical challenges that someone with more experience in how this works in the backend should assess.

If we follow your suggestion and rely on color's `name` and `value` only, we'll still need to factor in the edge-case scenario in the `ColorPalette` component where two separate color entries will share the same `name` and `color`. What do you think we should do, in that case? Behave like we do currently? Show only the first instance of that color, and hide the duplicate entries?

**Comment by jasmussen** [comment]
Ah, I meant that "Interface mode black" gives me `--wp--preset--color--interface-mode-black` as a variable, whereas "Black" gives me `--wp--preset--color--interface-mode-black`. I don't know if that's any better than the name, though, as it sounds like you could also have two identical CSS variable names here.

**Comment by ciampo** [comment]
When I refer to "color entries" having a `name` and a color `value` , I refer to the data format that the `ColorPalette` components expects to receive.

You can see it in action by visiting the [Storybook example](https://wordpress.github.io/gutenberg/?path=/docs/components-colorpalette--default) and clicking on the "Docs" tab in the top of the screen. There, you should see code snippet for different examples where the list of colors is an array of objects containing `name` and `value` — which means that it's all what the `ColorPalette` component can currently use to distinguish duplicate entries.

File: packages/components/src/color-palette/index.js [hunk 1]
```diff
@@ -2,7 +2,6 @@
 /**
  * External dependencies
  */
-import { map } from 'lodash';
```
Comment (mirka) [inline packages/components/src/color-palette/index.js:5]: 👋 Bye bye

**Review by mirka — APPROVED** [review]
I agree that this behavior is probably the lesser of all evils, in that the UI is honest about its current limitations. It's a bit confusing when you click Blue2 and Blue1 gets selected, but ultimately that is going to be the color slug you see when you reload the page.

So until we get backend changes that allow more than a raw color value to be persisted, I think this is a compromise we have to live with. (And the backend part should be discussed as a separate issue.)

**Review by mirka — CHANGES_REQUESTED** [review]
> To avoid two instances of the same color appearing as selected, I've arbitrarily decided that only the first matching color will appear as selected. This is not ideal, but IMO it's better than the current behavior.

I just realized we can't do this because it doesn't make sense when you're using a screen reader. When you select the second instance of the color, it sounds like the UI is unresponsive. Maybe we should keep this PR to the `key` fix only?

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for getting the discussion going on solutions to this issue @ciampo 👍 

@mirka makes a great point regarding the accessibility problem with the PR's current approach. +1 to keeping this PR limited to the `key` fix. Tackling the bigger problem should probably be explored alongside what changes might need to be made to the editors, block supports, theme.json, and global styles.

> (note that currently we're not keeping track of the name associated to the `color` that was clicked by the user)

I believe our editors actually attempt to map the color value back to a preset and store the slug in block attributes etc. For example, `textColor` or `backgroundColor`. This slug (name) is used to apply appropriate classnames, CSS var strings in the form of `var:preset|color|${ slug }`, and create the actual CSS vars themselves. Given that, duplicate slugs or names are likely also a problem in their own right.

Any solution to uniquely identifying color presets in the `ColorPalette` component will, in all likelihood, be required by our editors.

@andrewserong and @ramonjd will have more insight into how potentially storing additional color data might impact, or be impacted by, the style engine.

**Comment by ramonjd** [comment]
> This slug (name) is used to apply appropriate classnames, CSS var strings in the form of var:preset|color|${ slug }, and create the actual CSS vars themselves. Given that, duplicate slugs or names are likely also a problem in their own right.

> @andrewserong and @ramonjd will have more insight into how potentially storing additional color data might impact, or be impacted by, the style engine.

Thanks for the ping!

It depends on whether any additional data appears in a block's attributes and that data has any relationship to the block's actual text/background/gradient value. 

For presets, the style engine does expect a value format of `var:preset|color|${ slug }` as mentioned. In the backend we're constructing those from the attributes themselves. For example, see: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php#L99

Is there anywhere suggesting that block attribute values such as `backgroundColor` would change?

For all other purposes, the style engine mainly cares about the block attribute's `style` property, so if that model doesn't change we're golden. 

If there are multiple properties with the same value we'd dedupe them anyway.

Not sure if that even helps, but 🍺

**Comment by andrewserong** [comment]
> Thanks for getting the discussion going on solutions to this issue @ciampo 👍

+1 thanks for the discussion here, it's a pretty nuanced issue! The only other thing I'd add is to this comment:

> Entries with the same color value but a different name should be considered duplicates?

I'm probably just echoing what's already been discussed, but I think ideally these wouldn't be duplicates, so that someone selecting "Foreground" in a theme color can then switch to a different theme variation where that color is then swapped out for a different color value, versus selecting a color that shouldn't (or probably won't) change on a theme switch, like the named "Black" from the default palette. TwentyTwentyTwo's variations are probably a good example as "Foreground" is set to black in the "Default" variation but takes on a different color in each of the others.

**Comment by ciampo** [comment]
Thank you all for the replies! 

I'll go ahead and make this PR only focused on solving the `key` issue.

I will also open a new issue where the matter with duplicated color entries can be discussed separately

**Comment by ciampo** [comment]
> I will also open a new issue where the matter with duplicated color entries can be discussed separately

Opened #43197

**Review by mirka — APPROVED** [review]
Looks good! 🚢

**Pull request merged**